libutils/proc_manager.h

```diff
@@ -0,0 +1,123 @@
+/*
+  Copyright 2021 Northern.tech AS
+
+  This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+  This program is free software; you can redistribute it and/or modify it
+  under the terms of the GNU General Public License as published by the
+  Free Software Foundation; version 3.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#ifndef CFENGINE_PROC_MANAGER_H
+#define CFENGINE_PROC_MANAGER_H
+
+#include <sys/types.h>
+
+/**
+ * Struct to hold information about a subprocess.
+ */
+struct _Subprocess {
+    char *id;             /** unique ID of the subprocess                  [always not %NULL] */
+    char *cmd;            /** command used to create the subprocess        [can be %NULL] */
+    char *description;    /** human-friendly description of the subprocess [can be %NULL] */
+    pid_t pid;            /** PID of the subprocess                        [always >= 0] */
+    int in_fd;            /** FD of the input to the subprocess            [-1 if N/A] */
+    int out_fd;           /** FD of the output from the subprocess         [-1 if N/A] */
+    char lookup_fd;       /** which FD to use for lookup by FD             ['i', 'o' or '\0'] */
+};
+typedef struct _Subprocess Subprocess;
+
+/**
+ * Destroy a subprocess.
+ */
+void *SubprocessDestroy(Subprocess *proc);
+
+typedef struct _ProcManager ProcManager;
+
+/**
+ * Function to terminate a subprocess.
+ *
+ * @param subprocess_pid PID of the subprocess to terminate
+ * @param in_fd          FD of the input to the subprocess
+ * @param out_fd         FD of the output from the subprocess
+ * @param data           custom data passed to the function through the termination functions
+ * @return               whether the termination was successful or not
+ */
+typedef bool (*ProcessTerminator) (pid_t subprocess_pid, int in_fd, int out_fd, void *data);
+
+/**
+ * Get a new ProcManager.
+ */
+ProcManager *ProcManagerNew();
+
+/**
+ * Destroy a ProcManager.
+ */
+void *ProcManagerDestroy();
```

Comment:
Why is the return value of destroy `void*`?

Reply:
Good catch, should be `void` of course.